py/dml/c_backend.py

```diff
@@ -1691,26 +1693,38 @@ def eom_linemark(site):
     if dml.globals.linemarks:
         out('#line %d "%s"\n' % (site.lineno, quote_filename(site.filename())))
 
-def generate_each_in_table(node, trait, subnodes):
-    ident = EachIn.ident(node, trait)
-    items = []
-    for sub in subnodes:
-        # vtables exist, because trait instances are marked referenced
-        # when an 'each in' instance is referenced
-        ancestry_path = sub.traits.ancestry_paths[trait][0]
-        base = '&%s%s%s' % (
-            sub.traits.vtable_cname(ancestry_path[0]),
-            '[0]' * sub.dimensions,
-            ''.join('.' + cident(t.name)
-                    for t in ancestry_path[1:]))
-        num = reduce(operator.mul, sub.dimsizes, 1)
-        uniq = sub.uniq
-        offset = 'sizeof(struct _%s)' % (cident(ancestry_path[0].name),)
-        items.append("{(uintptr_t)%s, %d, %s, %d}\n"
-                     % (base, num, offset, uniq))
-    init = '{\n%s}' % (',\n'.join('    %s' % (item,) for item in items),)
-    add_variable_declaration(
-        'const _vtable_list_t %s[%d]' % (ident, len(subnodes)), init)
+def generate_each_in_tables():
+    tables = []
+
+    for ((node, trait), subnodes) in list(EachIn.instances.items()):
+        ident = EachIn.ident(node, trait)
+        items = []
+        for sub in subnodes:
+            # vtables exist, because trait instances are marked referenced
+            # when an 'each in' instance is referenced
+            ancestry_path = sub.traits.ancestry_paths[trait][0]
+            base = '&(*%s)%s%s' % (
+                sub.traits.vtable_cname(ancestry_path[0]),
+                '[0]' * sub.dimensions,
+                ''.join('.' + cident(t.name)
+                        for t in ancestry_path[1:]))
+            num = reduce(operator.mul, sub.dimsizes, 1)
+            uniq = sub.uniq
+            offset = 'sizeof(struct _%s)' % (cident(ancestry_path[0].name),)
+            items.append("{(uintptr_t)%s, %d, %s, %d}\n"
+                         % (base, num, offset, uniq))
+        add_variable_declaration(f'const _vtable_list_t *{ident}')
+        tables.append((ident, items))
+
+    start_function_definition('void _initialize_each_in_tables(void)')
+    out('{\n', postindent=1)
+    for (ident, items) in tables:
+        out(f'{ident} = MM_MALLOC({len(items)}, _vtable_list_t);\n')
```

Comment:
Does this really solve a problem? If you allocate mem and fill it with constants, then those constants still has to live somewhere; it doesn't get smaller than if you just put the whole initializer in bss.

And if huge size is an issue here, then the generated function will be huge, which is a problem for compile speed. We have previously seen that the time consumption of some of gcc's optimizations tend to be superlinear in function size (this is why initialize_traits was split into initialize_traitsN). Parsing initializers is instead extremely fast.

Reply:
The changes to the `each in` tables was because they now need to be initialized dynamically as the vtables memory addresses are no longer constant -- and at that point, I figured "why not also allocate them on the heap?"

I don't think the each in tables are a bottleneck, but if you think its best, I can implement code splitting up initialization into multiple functions just like we've done with trait vtables.

Reply:
Right, at least it's not subject to array explosion. Would it make sense to represent it as an array literal (wrapped in a struct), something like
```
void init_each_in_tables() {
  memcpy(x, &(const struct { _vtable_list_t _[N]; }){{{...}, {...}, ...}}, sizeof(x));
}
```
? or would this be too heavy on the stack?

Reply:
I'd say so. Naively, my current version should be as bad, as compound literals have the lifespan of the block, although I'm fairly sure GCC optimizes my current solution away. I shouldn't take the risk, though, and instead wrap each assignment in a block.

Reply:
Assignments updated to be wrapped in blocks. Question still remains if initialization should be split into multiple functions, from a discussion with Erik, we decided that should be decided from profiling.

Reply:
Profiling shows heavy regression on GCC compilation times on dml models with many large each-in tables due to dynamic initialization. Alternate solution implemented which allows each-in tables to keep being initialized statically.